XUtils.c

```diff
@@ -402,3 +402,83 @@ unsigned int countTrailingZeros(unsigned int x) {
    return mod37BitPosition[(-x & x) % 37];
 }
 #endif
+
+wchar_t* xMbstowcs(const char *mbs) {
+   size_t len = strlen(mbs);
+   wchar_t *wcs = xCalloc(len + 1, sizeof(wchar_t));
+   mbstate_t mbstate = {0};
+   mbsrtowcs(wcs, &mbs, len + 1, &mbstate);
+   if (mbs)
+      fail();
+   return wcs;
+}
+
+char* xWcstombs(const wchar_t *wcs) {
+   size_t len = wcslen(wcs);
+   size_t mbsSize = len * sizeof(wchar_t) + 1;
+   char *mbs = xMalloc(mbsSize);
+   mbstate_t mbstate = {0};
+   wcsrtombs(mbs, &wcs, mbsSize, &mbstate);
+   if (wcs)
+      fail();
+   return mbs;
+}
+
+wchar_t* xWcsdup(const wchar_t* str) {
+   wchar_t* data = wcsdup(str);
+   if (!data)
+      fail();
+   return data;
+}
+
+wchar_t* xWcsndup(const wchar_t* str, size_t len) {
+   wchar_t* data = xCalloc(len + 1, sizeof(wchar_t));
+   if (!data)
+      fail();
+
+   for (size_t i = 0; i < len && str[i]; i++)
+      data[i] = str[i];
```

Comment:
Use `memcpy` and write explicit terminator?

Reply:
I preferred `wcsncpy`, but I can change it to `memcpy` if you like.

Reply:
@onlined 
`wcsncpy` has an additional behavior of padding `L'\0'` to the unused character space. If you don't require this behavior, it's better to avoid it as it wastes processor cycles.

Reply:
`memcpy` would copy `len` bytes too, so should I go back to explicitly copying in a loop?

Reply:
@onlined The compiler can inline `memcpy` when it's useful. When it comes to performance, `memcpy` is arguably the best.

Reply:
Changed to `memcpy`.

Reply:
I just realized a problem, for using `memcpy` we have to assume that `str` has more characters than `len` or we can risk accessing unallocated memory and crash. Because of that concern, I changed it to use `Wstring_safeWcsncpy`.